### PR TITLE
call into snapd to set up encryption when required

### PR DESCRIPTION
**Pull request by mwhudson**

### CODE REVIEW

File: subiquity/server/controllers/install.py [hunk 1]
```diff
@@ -258,7 +259,19 @@ def make_curtin_step(name, stages, acquire_config, *,
             steps.extend([
```
Comment (ogayot) [inline subiquity/server/controllers/install.py:259]: steps.append() ?

**Review by ogayot — APPROVED** [review]
The logic seems good.  Just a few minor suggestions.

File: subiquity/models/filesystem.py [hunk 2]
```diff
@@ -958,6 +958,19 @@ def size(self):
         return self.volume.size - LUKS_OVERHEAD
 
 
+@fsobj("device")
+class ArbitraryDevice(_Device):
+    ptable = attr.ib(default=None)
```
Comment (dbungert) [inline subiquity/models/filesystem.py:963]: not that it's a new problem but we're overdue for a ptable enum

File: subiquity/models/filesystem.py [hunk 3]
```diff
@@ -1444,13 +1457,29 @@ def can_emit(obj):
                 raise Exception("\n".join(msg))
             work = next_work
 
+        if early:
+            r = [act for act in r if act['type'] not in ('format', 'mount')]
+        if late:
+            r = [act for act in r if act['type'] in ('format', 'mount')]
```
Comment (dbungert) [inline subiquity/models/filesystem.py:1463]: The connection between early & late & mounts is not immediately clear, I suggest a comment.
if early and later are both true, we'll drop the list from r, which I assume is always an error but would go unnoticed.

File: scripts/validate-yaml.py [hunk 4]
```diff
@@ -75,13 +75,17 @@ def final_checks(self):
...
 
 
-config = yaml.safe_load(open(sys.argv[1]))
 
 
 def main():
     storage_checker = StorageChecker()
 
-    for action in config['storage']['config']:
+    actions = []
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply scripts/validate-yaml.py:83]: well ok. I think Dan is going to make this script obsolete soon so I'll leave it for now :)

File: subiquity/server/controllers/install.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/server/controllers/install.py]: fair

File: subiquity/models/filesystem.py (hunk 2 above)
Comment (mwhudson) [reply subiquity/models/filesystem.py:964]: heh yes. not right now though!

File: subiquity/models/filesystem.py (hunk 3 above)
Comment (mwhudson) [reply subiquity/models/filesystem.py:1486]: I did a bit of refactoring here to make this all a bit clearer (basically _render_actions now takes an enum).

Comment (dbungert) [reply subiquity/models/filesystem.py:1486]: Thank you, this leaves less room for surprises for the caller.

**Review by dbungert — APPROVED** [review]

**Pull request merged**